packages/lodestar/src/api/impl/validator/utils.ts

```diff
@@ -43,25 +54,46 @@ export function computeSubnetForCommitteesAtSlot(
  * Precompute all pubkeys for given `validatorIndices`. Ensures that all `validatorIndices` are known
  * before doing other expensive logic.
  *
- * Note: Using a MutableVector is the fastest way of getting compressed pubkeys.
+ * Uses special BranchNodeStruct state.validators data structure to optimize getting pubkeys.
+ * Type-unsafe: assumes state.validators[i] is of BranchNodeStruct type.
+ * Note: This is the fastest way of getting compressed pubkeys.
  *       See benchmark -> packages/lodestar/test/perf/api/impl/validator/attester.test.ts
  */
 export function getPubkeysForIndices(
-  state: CachedBeaconState<allForks.BeaconState>,
-  validatorIndices: ValidatorIndex[]
-): (validatorIndex: ValidatorIndex) => BLSPubkey {
-  const validators = state.validators.persistent;
-  const pubkeyMap = new Map(
-    validatorIndices.map((validatorIndex) => {
-      const validator = validators.get(validatorIndex);
-      if (!validator) throw new ApiError(400, `Validator index ${validatorIndex} not in state`);
-      return [validatorIndex, validator.pubkey];
-    })
-  );
+  validators: allForks.BeaconState["validators"],
+  indexes: ValidatorIndex[]
+): BLSPubkey[] {
+  const validatorsLen = validators.length; // Get once, it's expensive
+  const validatorsTree = ((validators as unknown) as TreeValue<phase0.Validator>).tree;
```

Comment:
Not that its being used here, but the correct type is `TreeValue<List<phase0.Validator>>`, same below